src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Dependency/DependencyDag.cs

```diff
@@ -0,0 +1,797 @@
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Performance.SDK.Processing;
+using Microsoft.Performance.SDK.Runtime.Discovery;
+using Microsoft.Performance.SDK.Runtime.Extensibility.DataExtensions.DataCookers;
+using Microsoft.Performance.SDK.Runtime.Extensibility.DataExtensions.Repository;
+
+namespace Microsoft.Performance.SDK.Runtime.Extensibility.DataExtensions.Dependency
+{
+    /// <summary>
+    ///     Represents the complete graph of all Data Extensions
+    ///     and their dependencies as loaded in the system.
+    ///     <para/>
+    ///     Each node has a link to its dependencies and dependents.
+    ///     In the context of a given node, a _dependency_ is a node
+    ///     that is required for the given node, and a _dependent_ is
+    ///     a node that needs the current node. A node is a dependency
+    ///     of a dependent.
+    ///     <para/>
+    ///     Because this DAG represents dependency chains, we are going
+    ///     to define the "roots" to be all nodes that have no dependents. 
+    /// </summary>
+    public sealed class DependencyDag
+    {
+        private DependencyDag(IEnumerable<Node> allNodes)
+        {
+            Debug.Assert(allNodes != null);
+
+            this.All = new ReadOnlyHashSet<Node>(allNodes.ToSet());
+            this.Roots = new ReadOnlyHashSet<Node>(allNodes.Where(x => x.Dependents.Count == 0).ToSet());
+        }
+
+        /// <summary>
+        ///     Gets all <see cref="Node"/>s in the graph.
+        /// </summary>
+        public IReadOnlyCollection<Node> All { get; }
+
+        /// <summary>
+        ///     Gets all <see cref="Node"/>s that do not have
+        ///     any dependents.
+        /// </summary>
+        public IReadOnlyCollection<Node> Roots { get; }
+
+        /// <summary>
+        ///     Creates a new instance of the <see cref="DependencyDag"/>
+        ///     class from all of the loaded plugins / extensions.
+        /// </summary>
+        /// <param name="catalog">
+        ///     The catalog containing all of the loaded plugins.
+        /// </param>
+        /// <param name="repository">
+        ///     The repository containing all of the loaded extensions.
+        /// </param>
+        /// <returns>
+        ///     A new <see cref="DependencyDag"/> that is a graph of all
+        ///     of the dependencies between components.
+        /// </returns>
+        /// <exception cref="System.ArgumentException">
+        ///     The <paramref name="catalog"/> is not loaded.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="catalog"/> is <c>null</c>.
+        ///     - or -
+        ///     <paramref name="repository"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="System.InvalidOperationException">
+        ///     A dependency cycle exists between two or more
+        ///     componenets in the graph.
+        /// </exception>
```

Comment:
Probably need to add:
        /// <exception cref="ObjectDisposedException">
        ///     This instance is disposed.
        /// </exception>

since catalog.IsLoaded can throw this. I hate that this needs to be manually bubbled up. i feel like we're going to be missing this exception everywhere now

Reply:
Thanks, I'll add it.

Reply:
After looking at it more, I'll catch any ObjectDisposedExceptions and rethrow as InvalidOperation exceptions as it is not valid to compute the graph from disposed repositories.